iroh-net/src/hp/portmapper/mapping.rs

```diff
@@ -0,0 +1,233 @@
+}
+
+impl<M: Mapping> ActiveMapping<M> {
+    fn new(mapping: M) -> Self {
+        let deadline = Box::pin(time::sleep(mapping.half_lifetime()));
+        ActiveMapping {
+            mapping,
+            deadline,
+            expire_after: false,
+        }
+    }
+}
+
+/// Events in the lifetime of the mapping.
+#[derive(Debug, PartialEq, Eq)]
+pub(super) enum Event {
+    /// On this event, the mapping is halway through it's lifetime and should be renewed.
+    Renew { external_port: NonZeroU16 },
+    /// Mapping has expired.
+    Expired { external_port: NonZeroU16 },
+}
+
+/// Holds the current mapping value and ensures that any change is reported accordingly.
+#[derive(derive_more::Debug)]
+pub(super) struct CurrentMapping<M = super::upnp::Mapping> {
+    /// Active port mapping.
+    mapping: Option<ActiveMapping<M>>,
+    /// A [`watch::Sender`] that keeps the latest external address for subscribers to changes.
+    address_tx: watch::Sender<Option<SocketAddrV4>>,
+    /// Waker to ensure this is polled when needed.
+    #[debug(skip)]
+    waker: Option<std::task::Waker>,
+}
+
+impl<M: Mapping> CurrentMapping<M> {
+    /// Creates a new [`CurrentMapping`] and returns the watcher over it's external address.
+    pub(super) fn new() -> (Self, watch::Receiver<Option<SocketAddrV4>>) {
+        let (address_tx, address_rx) = watch::channel(None);
+        let wrapper = CurrentMapping {
+            mapping: None,
+            address_tx,
+            waker: None,
+        };
+        (wrapper, address_rx)
+    }
+
+    /// Updates the mapping, informing of any changes to the external address. The old mapping is
+    /// returned.
+    pub(super) fn update(&mut self, mapping: Option<M>) -> Option<M> {
+        trace!("New port mapping {mapping:?}");
+        let maybe_external_addr = mapping.as_ref().map(|mapping| {
+            let (ip, port) = mapping.external();
+            SocketAddrV4::new(ip, port.into())
+        });
+        let old_mapping = std::mem::replace(&mut self.mapping, mapping.map(ActiveMapping::new))
+            .map(|mapping| mapping.mapping);
+        // mapping changed
+        // TODO(@divma): maybe only wake if mapping is some
+        if let Some(waker) = &self.waker {
+            waker.wake_by_ref()
+        }
+        self.address_tx.send_if_modified(|old_addr| {
+            // replace the value always, as it could have different internal values
+            let old_addr = std::mem::replace(old_addr, maybe_external_addr);
+            // inform only if this produces a different external address
+            let update = old_addr != maybe_external_addr;
+            if update {
+                inc!(ExternalAddressUpdated);
+            };
+            update
+        });
+        old_mapping
+    }
+
+    fn poll(&mut self, cx: &mut std::task::Context<'_>) -> Poll<Event> {
+        // grab the waker if needed
+        if let Some(waker) = &self.waker {
+            if waker.will_wake(cx.waker()) {
```

Comment:
what exactly does this check do?

Reply:
an optimization to prevent switching wakers when the task has not moved threads